[Feature] Store config file by version

roycephan opened the pull request:
#### Status

READY

#### What's this PR do?

- [x] Generate checksum when sync new config
- [x] Read/Write config in persistence
- [x] Endpoint get list version
- [x] Endpoint revert version
- [x] Load config in persistent when run app

roycephan committed:
version control config

- persistence boltdb
- modify sync
- endpoint list ver, revert

roycephan committed:
seprate BoltPersistence into 3 interface

luquehuong committed:
feat: store new version into persistence

roycephan committed:
version control config

- endpoint revert version
- endpoint list version
- read/write config in boltDB

hieunmce commented on backend/config/boltdb.go:
```diff
@@ -0,0 +1,90 @@
+	"github.com/boltdb/bolt"
+)
+
+type BoltReaderWriterQuerierImpl struct {
```
We should make Impl private

hieunmce replied:
Also name of impl should be `boltImpl`

hieunmce commented on backend/config/boltdb.go:
```diff
@@ -0,0 +1,90 @@
+	db      *bolt.DB
+}
+
+func NewBoltReader(version string, db *bolt.DB) Reader {
```
we Should not duplicate New Function like this

hieunmce replied:
We can fix this by create an interface compose (reader, writer, Querier) and return that interface to prevent dupplicate

hieunmce commented on backend/endpoints/revert_version.go:
```diff
@@ -0,0 +1,45 @@
+		reader := config.NewBoltReader(req.Version, cfg.PersistenceDB)
+		vCfg, err := reader.Read()
+		if vCfg != nil {
+			vCfg.Lock()
```
Why we call Lock() in here

hieunmce commented on backend/config/boltdb.go:
```diff
@@ -0,0 +1,90 @@
+				return err
+			}
+		}
+		return bucket.Put([]byte(cfg.Version), buff)
```
We should add more meaning for cfg.Version.
example: `version_number:unix_timestamp:checksum`

hieunmce commented on backend/config/config.go:
```diff
@@ -57,11 +69,19 @@ func (c *Config) DB() *gorm.DB {
+	if err != nil {
+		return "", err
+	}
+	sum := sha256.Sum256(buff)
```
We should use MD5 for shorter key

hieunmce commented on backend/endpoints/list_version.go:
```diff
@@ -0,0 +1,24 @@
+	"github.com/go-kit/kit/endpoint"
+)
+
+type listVersionResponse struct {
```
ListVersion endpoint should have more meanful feature like version sequence name, timestamp of last sync..etc

hieunmce commented on backend/endpoints/list_version.go:
```diff
@@ -0,0 +1,24 @@
+import (
+	"context"
+
+	"github.com/dwarvesf/smithy/backend/config"
```
we should re-order import

hieunmce commented on backend/config/config.go:
```diff
@@ -82,18 +102,54 @@ func (c *Config) UpdateConfigFromAgent() error {
+		return nil
+	}
+	tempCfg.Version = checksum
+	tempCfg.Lock()
```
Why we lock without unlock (missing defer unlock?)

roycephan committed:
cooler code

- change sha256 to md5
- add more info to ListVersion response endpoint
- create interface compose (reader, writer, querier)

roycephan committed:
fix checksum

roycephan committed:
load config in persistent

- Get lastest version config
- define SyncPersistent func

hieunmce commented on backend/config/config.go:
```diff
@@ -19,6 +22,23 @@ type Reader interface {
+	LastestVersion() (*Config, error)
+}
+
+type ReaderWriterQuerier interface {
```
Check all warning about public function / interface should have comment

hieunmce commented on backend/config/config.go:
```diff
@@ -82,16 +119,52 @@ func (c *Config) UpdateConfigFromAgent() error {
-	c.ModelList = agentCfg.ModelList
-	err = c.UpdateDB()
+	// Copy config file into tempCfg
+	tempCfg := *c
```
We should not pass pointer of struct have a lock

hieunmce commented on backend/config/boltdb.go:
```diff
@@ -0,0 +1,102 @@
+	return err
+}
+
+func (b boltImpl) ListVersion() []Version {
```
We should return an error

roycephan committed:
- rename name
- check warning
- fix duplicate version number
